README.md
# SpeedFan Information Display

## Information

### Getting SpeedFan
SpeedFan is available at <http://www.almico.com/speedfan.php>

### Compatibility
This plugin has been tested with SpeedFan 4.44. It should work with any version of SpeedFan that outputs the log file in the same format as 4.44. The add-on will run on any platform using Kodi Krypton or later (although testing and active support is only done on the current production version of Kodi), but SpeedFan is a Windows only program.

## SpeedFan Setup
This readme is not meant to be a tutorial on how to setup SpeedFan or enable logging. The SpeedFan web site is available to help you with that. To get SpeedFanInfo to properly parse the log file, you do need to label the various SpeedFan items in a specific way.  SpeedFan Information Display knows about five kinds of log items: temperatures, fan speeds, voltages, fan speed percentages, and a catch all 'other.'

### Temperatures
This is the first tab in the SpeedFan configuration window (labeled Temperatures). Any item you are logging needs it's label to end with .temp (not case sensitive) for SpeedFanInfo to parse it. (i.e. CPU.temp).

### Fan Speeds
This is the second tab in the SpeedFan configuration window (labeled Fans). Any item you are logging needs it's label to end with .speed (not case sensitive) for SpeedFanInfo to parse it. (i.e. CPU Fan.speed)

### Voltages
This is the third tab in the SpeedFan configuration window (labeled Voltages). Any item you are logging needs it's label to end with .voltage (not case sensitive) for SpeedFanInfo to parse it. (i.e. +12V.voltage)

### Fan Speed Percentages
This is the fourth tab in the SpeedFan configuration window (labeled Speeds). Any item you are logging needs it's label to end with .percent (not case sensitive) for SpeedFanInfo to parse it. (i.e. CPU Fan.percent) A special note about fan speed percentages. They will only show on the SpeedFanInfo screen if you name the percentage the sam as the fan speeds (i.e. CPU.speed and CPU.percent). If there is no corresponding percentage for a given speed, only the speed will be displayed.

### Other
If you want any other log item to show up on the SpeedFan display, you can label it ending with .other.  These items will show up in the OTHER section exactly as stored in the log.

## Addon Options
After you install this plugin, go to the plugin settings. There are a few options to set.

### General
*Primary Log File Title:* (default empty)<br />
The name to display at the top of the information summary.  Useful if you're showing the logs from more than one machine.

*Log File Location:* (default empty)<br />
The folder where SpeedFan is storing the log file. By default this will likely be C:/Program Files/SpeedFan/

*Use Compact Display*<br />
When selected you get a smaller sidebar style display of readings rather than the fullscreen version. If you select this option you will get a slider to select the transparency level of the background. This is a nice option if you want to be able to peek at the information while watching other media.

*Log Temperature Scale*<br />
This should be set to match the temperature scale to which you have SpeedFan set. This setting just tells the plugin what to put after the integer value for the temperature. This add-on will not convert from Celcius to Farenheit (or vice versa).

*Window Update Interval (in seconds)*<br />
The number of seconds between updates to the screen. Please note that SpeedFan will sometimes get it's logging a little out of sync, so sometimes the fan speed and fan speed percentages don't actually match. This only happens when your fan speeds are fluctuating pretty rapidly, so if you wait through a couple of screen refreshes, the numbers usually start matching again.

### Additional Logs
*Use Log File x:* (default false)<br />
Enable one or both of these to monitor other SpeedFan logs. Once enabled, you need to provide a title and location for the log.

### Advanced
*Set Line Read Size:* (default 256)<br />
You shouldn't ever really need to change this unless you're using an external script to generate a copy cat log file (see below).  Changing this will change the amount of data read at one time by the add-on.

*Enable debug logging:* (default false)<br />
When enabled, if you have Kodi logging set the DEBUG you will get a very verbose set of information in your log file. You should only need to activate this when troubleshooting issues.

## Tips and Tricks

### Setting a Keyboard Shortcut
If you are using the compact display, you will probably want to assign a keyboard shortcut to launch SpeedFanInfo (and then map that to a key on your remote). To do that you need to add a custom key mapping. Information on key mapping is available at http://wiki.xbmc.org/index.php?title=Keymap Here's an example of assigning SpeedFanInfo to launch when you press F1.

```xml
    <keymap>
        <global>
            <keyboard>
                <f1>RunAddon(script.speedfaninfo)</f1>
            </keyboard>
        </global>
    </keymap>
```

## Getting Help
Please see the thread in the Kodi forums for assistance.

<https://forum.kodi.tv/showthread.php?tid=122481>

## Beta Testing
If you are interested in beta testing new versions of this add on (or just like being on the bleeding edge of up to date), you can install beta versions (Leia or later, there will be no more updates for earlier versions) from [my addon beta repository](https://github.com/pkscout/repository.beta.pkscout/raw/helix/repository.beta.pkscout-1.1.1.zip) (this is a zip download). You can also monitor the support thread, as links to particular beta releases will be available there as well.

CHANGELOG.txt
v.1.0.1~beta1
- changed XBMC to Kodi in addon description

v.1.0.0
- Python 2/3 compatibility
- converted language files to current format for forward compatability
- fix for addon not quitting properly if running when Kodi attempts to exit
- updated for new method of retrieving preferences
- updated icon

v.0.7.8
- fix for problem with common library moving files
- changes for better compatibility when running script on non-Windows machine

v.0.7.7
- fix for crash if launching external script

v.0.7.6
- updated common libraries (including fix for subprocess not available in Windows Store version of Kodi)

v.0.7.5
- added option to run external script to generate log file
- added 'Other' option to sensor display
- added advanced setting for line read size

v.0.7.4
- internal release

v.0.7.3
- convert Celcius in Speedfan log to Fahrenheit when Fahrenheit is selected in settings

v.0.7.2
- fix for incorrect display in fullscreen

v.0.7.1
- updated for Helix
- fix for display issue in Kodi v14 (Helix)

v.0.6.1
- if rpm reported as zero 0, then SFID only shows percentage

v.0.6.0
- fix for crash when using smb path

v.0.5.9
- fix for required xbmc python version
- fix for malformed setting.xml file
- included XML encoding declaration in skin files

v.0.5.8
- suppress all logging by default with option to turn debug logging on
- updated language files

v.0.5.7
- added full support for logfile on remote file system without mapping drive in OS
- fixed crash when log file can't be parsed
- updated language files

v.0.5.6
- fixed crash when logging non-text objects

v.0.5.5
- added ability to monitor multiple logs
- script now sets focus on window so you can scroll through it without mouse control
- added additional metadata to addon
- updated module import to follow best practice
- cleaned up code and moved some stuff into external modules

v 0.5.4
- changed display logic so percents will show even if the number of percent entries doesn't match number of speed entries
- added Galician language files
- changed over to Frodo .po language files
- added check to threading loop for XBMC quitting
- updated addon.xml with new information and format for repo database
- updated logging so it uses standard XBMC logging levels

v.0.5.3
- internal testing version

v.0.5.2
- updated language files from Transifex
- added compact display option
- added logic to keep script from being run multiple times

v0.5.1
- internal testing version

v0.5.0
- updated language files from Transifex

v.0.3.5
-changed the way the degree sign is displayed as old way didn't work either on OSX or with Frodo (not sure which)
-included new language files for XBMC Frodo and added addon to transifex translation engine

v.0.3.4
-added support for languages that use comma as the ordinal separator for decimals

v.0.3.3
-First public release
-Added visual notice if there is no log file location defined or no log file in the defined location
-Fixed a typo in one of the verbose logging outputs
-added a README file with instructions on how to setup the SpeedFan log file

v.0.3.2
-Fixed issue with voltage headings not displaying correctly.  Required a change in the way you format the SpeedFan log, but since nobody is using it at this point but me, it doesn't matter. '-)
-Fixed the way the setting.xml passes the temperature scale to default.py so that the code is a little easier to read.

v.0.3.1
-Fixes to logging engine
-Updated display XML to be skin independent (mostly having to do with font settings)
-Moved all text strings into language file so that the plugin can be localized
-Added icon

v.0.3.0
-Implemented new logging engine to make it easier to have a basic log for day to day use and full log for troubleshooting
-updated code to make better use of boolean objects

v.0.2.0
-Added ability to update screen based on arbitary interval (set by user)

v.0.1.1
-Fixed some code issues that caused the script not to work on Windows (that's what you get when you develop on a Mac and deploy to Windows)

v.0.1.0
-Initial internal release

default.py
from kodi_six import xbmc, xbmcaddon, xbmcgui
from threading import Thread
from resources.common.kodisettings import getSettingBool, getSettingString
from resources.common.xlogger import Logger
from resources.lib.spid import Main, updateWindow

addon        = xbmcaddon.Addon()
addonversion = addon.getAddonInfo( 'version' )
addonpath    = addon.getAddonInfo( 'path' )
preamble     = '[SpeedFan Info]'
logdebug     = addon.getSetting( 'logging' )

lw = Logger( preamble = preamble, logdebug = logdebug )

if ( __name__ == "__main__" ):
    lw.log( ['script version %s started' % addonversion], xbmc.LOGNOTICE )
    lw.log( ['debug logging set to %s' % logdebug], xbmc.LOGNOTICE )
    if xbmcgui.Window( 10000 ).getProperty( "speedfan.running" ) == "true":
        lw.log( ['script already running, aborting subsequent run attempts'] )
    else:
        xbmcgui.Window( 10000 ).setProperty( "speedfan.running",  "true" )
        if getSettingBool( addon, 'show_compact' ):
            transparency_image = 'speedfan-panel-compact-%s.png' % getSettingString( addon, 'transparency', default='70' )
            xbmcgui.Window( 10000 ).setProperty( 'speedfan.panel.compact',  transparency_image )
            w = Main( "speedfaninfo-compact.xml", addonpath )
        else:
            w = Main( "speedfaninfo-main.xml", addonpath )
        t1 = Thread( target=updateWindow,args=("thread 1", w) )
        t1.setDaemon( True )
        t1.start()
        w.doModal()
        del t1
        del w
        xbmcgui.Window(10000).setProperty( "speedfan.running",  "false" )
lw.log( ['script stopped'], xbmc.LOGNOTICE )

resources/common/url.py
import requests
    
__timeout__    = 10
__headers__    = ''
__returntype__ = 'text'


class URL():
    def __init__( self, returntype='text', headers='', timeout=10 ):
        global __timeout__, __headers__, __returntype__
        __timeout__ = timeout
        __headers__ = headers
        __returntype__ = returntype

    
    def Get( self, url, **kwargs ):
        params, data = self._unpack_args( kwargs )
        return self._urlcall( url, params, '', 'get' )
    
    
    def Post( self, url, **kwargs ):
        params, data = self._unpack_args( kwargs )
        return self._urlcall( url, params, data, 'post' ) 
    
    
    def Delete( self, url, **kwargs ):
        params, data = self._unpack_args( kwargs )
        return self._urlcall( url, params, data, 'delete' )
    
    
    def _urlcall( self, url, params, data, urltype ):
        loglines = []
        urldata = ''
        try:
            if urltype == "get":
                urldata = requests.get( url, params=params, timeout=__timeout__ )
            elif urltype == "post":
                urldata = requests.post( url, params=params, data=data, headers=__headers__, timeout=__timeout__ )
            elif urltype == "delete":
                urldata = requests.delete( url, params=params, data=data, headers=__headers__, timeout=__timeout__ )
        except requests.exceptions.ConnectionError, e:
            loglines.append( 'site unreachable at ' + url )
            loglines.append( e )
        except requests.exceptions.Timeout, e:
            loglines.append( 'timeout error while downloading from ' + url )
            loglines.append( e )
        except requests.exceptions.HTTPError, e:
            loglines.append( 'HTTP Error while downloading from ' + url )
            loglines.append( e )
        except requests.exceptions.RequestException, e:
            loglines.append( 'unknown error while downloading from ' + url )
            loglines.append( e )
        if urldata:
            success = True
            if __returntype__ == 'text':
                data = urldata.text()
            elif __returntype__ == 'binary':
                data = urldata.content()
            elif __returntype__ == 'json':
                data = urldata.json()
                if data == None:
                    data = []
        else:
            success = False
            data = ''
        loglines.append( '-----URL OBJECT RETURNED-----' )
        loglines.append( data )
        return success, loglines, data
    
    
    def _unpack_args( self, kwargs ):
        try:
            params = kwargs['params']
        except:
            params = {}
        try:
            data = kwargs['data']
        except:
            if __returntype__ == 'json':
                data = []
            else:
                data = ''
    	return params, data

resources/xlogger/xlogger.py
import xbmc

__log_preamble__ = ''

#this class creates an object used to log stuff to the xbmc log file
class Logger():
    def __init__(self): pass
        # and define it as self

    def setPreamble(self, preamble):
        #sets the preamble for the log line so you can find it in the XBMC log
        global __log_preamble__
        __log_preamble__ = preamble

    def parseListorTuple(self, items, count):
        #parse a list or tuple into a string
        #this function will recurse if it detects lists of lists of lists, etc.
        #define a whole bunch of possible delimiters for recursion purposes
        delims = [',', ':', '/', ';', '~', '!', '@', '#', '$', '%', '^', '&', '*']
        line = ''
        for item in items:
            itemclass = item.__class__.__name__
            if(itemclass == 'list' or itemclass == 'tuple'):
                grouping = self.parseListorTuple(item, count+1)
            else:
                grouping = item
            if(len(line) == 0):
                line = grouping
            else:
                line = line + delims[count] + ' ' + grouping
        return line
        
    def log(self, *args):
        #send an arbitrary group of data to log
        #the last argument must be the logtype of the data (i.e. standard or verbose)
        #convert
        #l_args = list(args)
        #the log_level is in the last item of the tuple
        log_level = args[-1]
        #now we need to iterate through all the other args and log them
        for arg in args[:-1]:
            argclass = arg.__class__.__name__
            if(argclass == 'str' or argclass == 'unicode'):
                line = arg
            elif(argclass == 'list' or argclass == 'tuple'):
                #loop through the items and put them into comma separated list
                line = self.parseListorTuple(arg, 0)
            else:
                line = 'no appropriate action found for class ' + argclass
            xbmc.log(__log_preamble__ + ' ' + line, log_level)


resources/lib/spid.py
import datetime, os
from kodi_six import xbmc, xbmcaddon, xbmcgui, xbmcvfs
try:
    from itertools import izip_longest as _zip_longest
except ImportError:
    from itertools import zip_longest as _zip_longest
from resources.lib.xlogger import Logger
from resources.lib.kodisettings import getSettingBool, getSettingInt, getSettingString

addon        = xbmcaddon.Addon()
addonname    = addon.getAddonInfo( 'id' )
addonversion = addon.getAddonInfo( 'version' )
addonpath    = addon.getAddonInfo( 'path' )
addonicon    = xbmc.translatePath( '%s/icon.png' % addonpath )
language     = addon.getLocalizedString
preamble     = '[SpeedFan Info]'
logdebug     = getSettingBool( addon, 'logging', default=False )

lw = Logger( preamble=preamble, logdebug=logdebug )

lw.log( ['script version %s started' % addonversion], xbmc.LOGNOTICE )
lw.log( ['debug logging set to %s' % logdebug], xbmc.LOGNOTICE )



class SpeedFanWindow( xbmcgui.WindowXMLDialog ):

    def __init__( self, *args, **kwargs ):
        self.KEEPRUNNING = True
        self.ACTION_PREVIOUS_MENU = 10
        self.ACTION_BACK = 92


    def onAction( self, action ):
        if action == self.ACTION_PREVIOUS_MENU or action == self.ACTION_BACK:
            lw.log( ['user initiated previous menu or back, telling addon to quit'] )
            self.KEEPRUNNING = False


    def KeepRunning( self ):
        return self.KEEPRUNNING



class Main():

    def __init__( self ):
        self._get_settings()
        self._init_vars()
        if self.GUIWINDOW.getProperty( 'SpeedFan.Running' ) == "True":
            lw.log( ['script already running, aborting subsequent run attempts'], xbmc.LOGNOTICE )
            return
        self.GUIWINDOW.setProperty( 'SpeedFan.Running',  'True' )
        self._init_window()
        self._populate_from_all_logs()
        if self.FOUNDLOGFILE:
            while not xbmc.Monitor().abortRequested() and self.SPEEDFANWINDOW.KeepRunning():
                for i in range( self.DELAY ):
                    if xbmc.Monitor().waitForAbort( 1 ) or not self.SPEEDFANWINDOW.KeepRunning():
                        break
                self._populate_from_all_logs()
        self.SPEEDFANWINDOW.close()
        self.GUIWINDOW.setProperty( 'SpeedFan.Running',  'False' )


    def _get_log_files( self ):
        log_file_date = datetime.date(2011,1,29).today().isoformat().replace('-','')
        log_files = []
        for info_set in self.LOGINFO:
            if info_set['use_log'] == 'true':
                log_file = os.path.join( info_set['loc'], 'SFLog' + log_file_date + '.csv' )
                log_files.append( (info_set['title'], log_file) )
        return log_files


    def _get_settings( self ):
        self.SHOWCOMPACT = getSettingBool( addon, 'show_compact' )
        self.TRANSPARENCY = getSettingString( addon, 'transparency', default='70' )
        self.TEMPSCALE = getSettingString( addon, 'temp_scale' )
        self.READSIZE = getSettingInt( addon, 'read_size' )
        self.DELAY = getSettingInt( addon, 'update_delay', default=30 )
        self.LOGINFO = []
        for i in range( 3 ):
            log_info = {}
            if i == 0:
                log_num = ''
                log_info['use_log'] = 'true'
            else:
                log_num = str( i + 1 )
                log_info['use_log'] = getSettingString( addon, 'use_log' + log_num )
            log_info['loc'] = getSettingString( addon, 'log_location' + log_num )
            log_info['title'] = getSettingString( addon, 'log_title' + log_num )
            self.LOGINFO.append( log_info )


    def _init_vars( self ):
        self.GUIWINDOW = xbmcgui.Window( 10000 )
        self.FOUNDLOGFILE = False


    def _init_window( self ):
        if self.SHOWCOMPACT:
            transparency_image = 'speedfan-panel-compact-%s.png' %  self.TRANSPARENCY
            self.GUIWINDOW.setProperty( 'speedfan.panel.compact',  transparency_image )
            self.SPEEDFANWINDOW = SpeedFanWindow( 'speedfaninfo-compact.xml', addonpath )
        else:
            self.SPEEDFANWINDOW = SpeedFanWindow( "speedfaninfo-main.xml", addonpath )
        self.SPEEDFANWINDOW.show()
        self.LISTCONTROL = self.SPEEDFANWINDOW.getControl( 120 )


    def _parse_log( self ):
        #parse the log for information, see readme for how to setup SpeedFan output so that the script
        lw.log( ['started parsing log'] );
        lw.log( ['read the log file'] )
        first, last = self._read_log_file()
        temps = []
        speeds = []
        voltages = []
        percents = []
        others = []
        if first == '' or last == '':
            return temps, speeds, voltages, percents, others
        #pair up the heading with the value
        lw.log( ['pair up the heading with the value'] );
        for s_item, s_value in _zip_longest( first.split( '\t' ), last.split( '\t' ) ):
            item_type = s_item.split( '.' )[-1].rstrip().lower()
            item_text = os.path.splitext( s_item )[0].rstrip()
            #round the number, drop the decimal and then covert to a string
            #skip the rounding for the voltage and other readings
            if item_type == 'voltage' or item_type == 'other':
                s_value = s_value.rstrip()
            else:
                try:
                    s_value = str( int( round( float( s_value.rstrip() ) ) ) )
                except ValueError:
                    s_value = str( int( round( float( s_value.rstrip().replace(',', '.') ) ) ) )
            if item_type == "temp":
                lw.log( ['put the information in the temperature array'] )
                if self.TEMPSCALE == 'Celcius':
                    temps.append( [item_text + ':', s_value + 'C'] )
                else:
                    temps.append( [item_text + ':', str( int( round( ( float( s_value ) * 1.8 ) + 32 ) ) ) + 'F'] )
            elif item_type == "speed":
                lw.log( ['put the information in the speed array'] )
                speeds.append( [item_text + ':', s_value + 'rpm'] )
            elif item_type == "voltage":
                lw.log( ['put the information in the voltage array'] )
                voltages.append( [item_text + ':', s_value + 'v'] )
            elif item_type == "percent":
                lw.log( ['put the information in the percent array'] );
                percents.append( [item_text, s_value + '%'] )
            elif item_type == "other":
                lw.log( ['put the information in the other array'] );
                others.append( [item_text + ":", s_value] )
        lw.log( [temps, speeds, voltages, percents, others, 'ended parsing log, displaying results'] )
        return temps, speeds, voltages, percents, others


    def _populate_from_all_logs( self ):
        lw.log( ['reset the window to prep it for data'] )
        self.LISTCONTROL.reset()
        displayed_log = False
        for title, logfile in self._get_log_files():
            self.LOGFILE = logfile
            if title:
                item = xbmcgui.ListItem( label=title )
                item.setProperty( 'istitle','true' )
                self.LISTCONTROL.addItem( item )
            if xbmcvfs.exists( logfile ):
                displayed_log = True
                self._populate_from_log()
        if displayed_log:
            self.SPEEDFANWINDOW.setFocus( self.LISTCONTROL )
            self.FOUNDLOGFILE = True
        else:
            lw.log( ['no logfile found, put up a notification and quitting'] )
            xbmcgui.Dialog().notification( language( 30103 ), language( 30104 ), icon=addonicon, time=6000 )


    def _populate_from_log( self ):
        #get all this stuff into list info items for the window
        temps, speeds, voltages, percents, others = self._parse_log()
        lw.log( ['starting to convert output for window'] )
        #add a fancy degree symbol to the temperatures
        formatted_temps = []
        for temp in temps:
            formatted_temps.append( [temp[0], temp[1][:-1] + u'\N{DEGREE SIGN}' + temp[1][-1:]] )
        #now parse all the data and get it into ListIems for display on the page
        #this allows for a line space *after* the first one so the page looks pretty
        firstline_shown = False
        lw.log( ['put in all the temperature information'] )
        if formatted_temps:
            self._populate_list( language( 30100 ), formatted_temps, firstline_shown )
            firstline_shown = True
        lw.log( ['put in all the speed information (including percentages)'] )
        if speeds:
            lw.log( ['adding the percentages to the end of the speeds'] )
            en_speeds = []
            for thespeed in speeds:
                #if there is a matching percentage, add it to the end of the speed
                percent_match = False
                percent_value = ''
                for thepercent in percents:
                    if (thespeed[0][:-1] == thepercent[0]):
                        lw.log( ['matched speed ' + thespeed[0][:-1] + ' with percent ' + thepercent[0]] )
                        percent_match = True
                        percent_value = thepercent[1]
                if percent_match:
                    if thespeed[1] == '0rpm':
                        en_speeds.append( (thespeed[0], percent_value) )
                    else:
                        en_speeds.append( (thespeed[0], thespeed[1] + ' (' + percent_value + ')') )
                else:
                    en_speeds.append( (thespeed[0], thespeed[1]) )
            self._populate_list( language( 30101 ), en_speeds, firstline_shown )
            firstline_shown = True
        lw.log( ['put in all the voltage information'] )
        if voltages:
            self._populate_list( language( 30102 ), voltages, firstline_shown )
        lw.log( ['put in all the other information'] )
        if others:
            self._populate_list( language( 30105 ), others, firstline_shown )
        #add empty line at end in case there's another log file
        item = xbmcgui.ListItem()
        self.LISTCONTROL.addItem( item ) #this adds an empty line
        lw.log( ['completed putting information into lists, displaying window'] )


    def _populate_list( self, title, things, titlespace ):
        #this takes an arbitrary list of readings and gets them into the ListItems
        lw.log( ['create the list item for the title of the section'] )
        if titlespace:
            item = xbmcgui.ListItem()
            self.LISTCONTROL.addItem( item ) #this adds an empty line
        item = xbmcgui.ListItem( label=title )
        item.setProperty( 'istitle','true' )
        self.LISTCONTROL.addItem( item )
        #now add all the data (we want two columns in full mode and one column for compact)
        if self.SHOWCOMPACT:
            lw.log( ['add all the data to the one column format'] )
            for onething in things:
                    item = xbmcgui.ListItem( label=onething[0],label2='' )
                    item.setProperty( 'value',onething[1] )
                    self.LISTCONTROL.addItem( item )
        else:
            lw.log( ['add all the data to the two column format'] )
            nextside = 'left'
            for  onething in things:
                if(nextside == 'left'):
                    left_label = onething[0]
                    left_value = onething[1]
                    nextside = 'right'
                else:
                    item = xbmcgui.ListItem( label=left_label,label2=onething[0] )
                    item.setProperty( 'value',left_value )
                    item.setProperty( 'value2',onething[1] )
                    nextside = 'left'
                    self.LISTCONTROL.addItem( item )
            if(nextside == 'right'):
                item = xbmcgui.ListItem( label=left_label,label2='' )
                item.setProperty( 'value',left_value )
                self.LISTCONTROL.addItem( item )


    def _parse_line( self, f, s_pos ):
        lw.log( ['parsing line'] )
        file_size = f.size()
        read_size = self.READSIZE
        if s_pos == 2:
            direction = -1
            offset = read_size
        else:
            direction = 1
            offset = 0
        while True:
            if file_size < read_size:
                read_size = file_size
            f.seek( direction*offset, s_pos )
            read_str = f.read( read_size )
            lw.log( [read_str] )
            # Remove newline at the end
            try:
                if read_str[offset - 1] == '\n':
                    read_str = read_str[0:-1]
            except IndexError:
                pass
            lines = read_str.split('\n')
            if len( lines ) > 1:  # Got a complete line
                if s_pos == 2:
                    return lines[-1]
                else:
                    return lines[0]
            if offset == file_size:   # Reached the beginning
                return ''
            offset += read_size


    def _read_log_file( self ):
        #try and open the log file
        lw.log( ['trying to open logfile ' + self.LOGFILE] )
        try:
            f = xbmcvfs.File( self.LOGFILE )
        except Exception as e:
            lw.log( ['unexpected error when reading log file', e], xbmc.LOGERROR )
            return ('', '')
        lw.log( ['opened logfile ' + self.LOGFILE] )
        #get the first and last line of the log file
        #the first line has the header information, and the last line has the last log entry
        first = self._parse_line( f, 0 )
        last = self._parse_line( f, 2 )
        f.close()
        lw.log( ['first line: ' + first, 'last line: ' + last] )
        return first, last